clippy_lints/src/needless_traits_in_scope.rs

```diff
@@ -0,0 +1,76 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use rustc_errors::Applicability;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Find traits that are not explicitely used in scope (like `AsRef::as_ref()`)
+    /// but directly with the trait method (like `opt.as_ref()`).
+    ///
+    /// These traits can be imported anonymously with `use crate::Trait as _`.
+    /// This avoids name collision with other traits (possibly with the same name).
+    /// It also helps identify the traits in `use` statements.
+    ///
+    /// ### Why is this bad?
+    /// This needlessly brings a trait in trait's namespace.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::io::Read;
+    /// fn main() {
+    ///   let mut b = "I'm your father!".as_bytes();
+    ///   let mut buffer = [0; 10];
+    ///   b.read(&mut buffer)
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::io::Read as _;
+    /// fn main() {
+    ///   let mut b = "I'm your father!".as_bytes();
+    ///   let mut buffer = [0; 10];
+    ///   b.read(&mut buffer)
+    /// }
+    /// ```
+    #[clippy::version = "1.69.0"]
+    pub NEEDLESS_TRAITS_IN_SCOPE,
+    pedantic,
```

Comment:
I think this should be a restriction lint. It's nothing that should be frowned upon in normal Rust code and makes the code longer, but it could bring benefits in certain situations (those you mentioned).

Reply:
Indeed, better than `pedantic`. Fixed in 02db9854e.